keanu-project/src/test/java/io/improbable/keanu/vertices/bool/probabilistic/ProbabilisticBooleanVertexTest.java

```diff
@@ -0,0 +1,72 @@
+package io.improbable.keanu.vertices.bool.probabilistic;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import static io.improbable.keanu.tensor.TensorMatchers.tensorEqualTo;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import org.junit.Test;
+
+import com.google.common.primitives.Booleans;
+
+import io.improbable.keanu.tensor.TensorShape;
+import io.improbable.keanu.tensor.bool.BooleanTensor;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+
+public class ProbabilisticBooleanVertexTest {
+    @Test
+    public void sampleUsesShapeOfTensor() {
+        ProbabilisticBooleanVertex vertex1 = new ProbabilisticBooleanVertexImplementation(1, 1);
+
+        assertThat(vertex1.sample(), tensorEqualTo(BooleanTensor.scalar(true)));
+
+        ProbabilisticBooleanVertex vertex2 = new ProbabilisticBooleanVertexImplementation(2, 3);
+
+        assertThat(vertex2.sample(), tensorEqualTo(BooleanTensor.create(true, false, true, false, true, false).reshape(2, 3)));
+    }
+
+    @Test
+    public void sampleManyScalarsUsesInputShape() {
+        ProbabilisticBooleanVertex vertex = new ProbabilisticBooleanVertexImplementation(1, 1);
+
+        assertThat(vertex.sampleManyScalars(new long[]{1, 1}), tensorEqualTo(BooleanTensor.scalar(true)));
+        assertThat(vertex.sampleManyScalars(new long[]{2, 3, 2}), tensorEqualTo(BooleanTensor.create(true, false, true, false, true, false, true, false, true, false, true, false).reshape(2, 3, 2)));
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void cannotSampleManyScalarsOfNonScalar() {
+        ProbabilisticBooleanVertex vertex = new ProbabilisticBooleanVertexImplementation(2, 2);
+
+        vertex.sampleManyScalars(new long[]{1, 1});
+    }
+
+    private static class ProbabilisticBooleanVertexImplementation extends ProbabilisticBooleanVertex {
+        ProbabilisticBooleanVertexImplementation(long... shape) {
+            this.setValue(BooleanTensor.placeHolder(shape));
+        }
+
+        @Override
+        protected BooleanTensor sampleWithShape(long[] shape, KeanuRandom random) {
```

Comment:
It doesn't really make sense to implement the concrete class `ProbabilisticBooleanVertexImplementation` - your tests are then dependent on its logic, rather than focussing on the logic of `ProbabilisticBooleanVertex`.

What you could do instead is instantiate a `mock(ProbabilisticBooleanVertex.class)` and then verify that its `sampleWithShape` method is called with the appropriate `shape` argument.

Reply:
Ditto for the Integer and Double versions.